testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/jca/capacitypolicies/AbstractDatasourceCapacityPoliciesTestCase.java

```diff
@@ -0,0 +1,306 @@
+        connections[0] = ds.getConnection();
+
+        // sometimes InUseCount is 2 and AvailableCount is 3 when statistics are checked right after
+        // ds.getConnection, hence this sleep. I guess it's caused by CapacityFiller
+        Thread.sleep(50);
+
+        checkStatistics(4, 1, 5, 0);
+
+        connections[1] = ds.getConnection();
+        checkStatistics(3, 2, 5, 0);
+
+        connections[2] = ds.getConnection();
+        checkStatistics(2, 3, 5, 0);
+
+        connections[3] = ds.getConnection();
+        checkStatistics(1, 4, 5, 0);
+
+        for (int i = 0; i < 4; i++) {
+            Connection c = connections[i];
+            c.close();
+        }
+
+        WrapperDataSource wsds = JcaTestsUtil.extractWrapperDatasource((WildFlyDataSource) ds);
+        ManagedConnectionPool mcp = JcaTestsUtil.extractManagedConnectionPool(wsds);
+        JcaTestsUtil.callRemoveIdleConnections(mcp);
+
+        checkStatistics(5, 0, 2, 3);
+    }
+
+    private void checkStatistics(int expectedAvailableCount, int expectedInUseCount,
+                                 int expectedActiveCount, int expectedDestroyedCount) throws Exception {
+        ModelNode statsAddress = xaDatasource ? XA_DS_ADDRESS.clone() : DS_ADDRESS.clone();
+        statsAddress.add("statistics", "pool");
+
+        int availableCount = readAttribute(statsAddress, "AvailableCount").asInt();
+        int inUseCount = readAttribute(statsAddress, "InUseCount").asInt();
+        int activeCount = readAttribute(statsAddress, "ActiveCount").asInt();
+        int destroyedCount = readAttribute(statsAddress, "DestroyedCount").asInt();
+
+        Assert.assertEquals("Unexpected AvailableCount", expectedAvailableCount, availableCount);
+        Assert.assertEquals("Unexpected InUseCount", expectedInUseCount, inUseCount);
+        Assert.assertEquals("Unexpected ActiveCount", expectedActiveCount, activeCount);
+        Assert.assertEquals("Unexpected DestroyedCount", expectedDestroyedCount, destroyedCount);
+    }
+
+    static abstract class AbstractDatasourceCapacityPoliciesServerSetup extends JcaMgmtServerSetupTask {
+        private boolean xa;
+
+        public AbstractDatasourceCapacityPoliciesServerSetup(boolean xa) {
+            this.xa = xa;
+        }
+
+        @Override
+        public void doSetup(final ManagementClient managementClient) throws Exception {
+            CapacityConfiguration configuration = new CapacityConfiguration(
+                    "org.jboss.jca.core.connectionmanager.pool.capacity.MinPoolSizeDecrementer",
+                    "org.jboss.jca.core.connectionmanager.pool.capacity.MaxPoolSizeIncrementer");
+            createDatasource(configuration);
+
+        }
+
+        @Override
+        public void tearDown(final ManagementClient managementClient, final String containerId) throws Exception {
+            removeDatasource();
+        }
+
+        private void createDatasource(CapacityConfiguration capacityConfiguration) throws Exception {
```

Comment:
I'm not sure if I understand, do you mean to move code from createDatasource directly do doSetup? (other tests add datasource in doSetup)